apps/lib/app_x509.c

```diff
@@ -107,8 +107,13 @@ static int do_x509_ctrl_string(int (*ctrl)(void *object, int cmd,
 #endif
     } else if (strcmp(stmp, "hexdistid") == 0) {
         long hexid_len = 0;
-        void *hexid = OPENSSL_hexstr2buf((const char *)vtmp, &hexid_len);
+        void *hexid;
 
+        if (vtmp == NULL) {
+            OPENSSL_free(stmp);
+            return -1;
+        }
+        hexid = OPENSSL_hexstr2buf((const char *)vtmp, &hexid_len);
```

Comment:
Alternately, if vtmp is NULL, hexid could be the empty string, let ctrl deal with it...

Reply:
Much of a muchness.

I've changed the code over to pass a null string instead.

Reply:
The alternative causes memory leaks in the tests :(

Reply:
Er, so deal with it?  Do you want me to take over this case?

Reply:
Is supplying an argument `hexdistid` without the colon or hex string really that meaningful?

Reply:
Something is wrong here.  The `sm2_pmeth` function that handles this repeated the `OPENSSL_hexstr2buf()` call but using the converted string.

Reply:
Hmmm, perhaps not.  I had to look around how we do this elsewhere to be sure...

Okie then

Reply:
> Something is wrong here. The `sm2_pmeth` function that handles this repeated the `OPENSSL_hexstr2buf()` call but using the converted string.

Huh...  let me have a look

Reply:
> Something is wrong here.  The `sm2_pmeth` function that handles this repeated the `OPENSSL_hexstr2buf()` call but using the converted string.

I don't understand how this happens...  it should not be called at all?

Reply:
The solution seems to be to pass `NULL` not `""`.

Reply:
I deleted my mistaken comment...